nidx/src/maintenance/scheduler.rs

```diff
@@ -167,37 +170,150 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
+            bottom_bucket_threshold: 1000,
+            bucket_size_log: 0.75,
         }
     }
+}
 
-    Ok(())
+impl MergeScheduler {
+    pub fn from_settings(settings: &MergeSettings) -> Self {
+        Self {
+            min_number_of_segments: settings.min_number_of_segments,
+            top_bucket_max_records: settings.max_segment_size,
+            ..Default::default()
+        }
+    }
+
+    /// Enqueue merge jobs for segments older than `last_indexed_seq` that aren't
+    /// already scheduled for merge or marked to delete.
+    ///
+    /// Merging involves creation of a single segment from multiple ones, combining
+    /// their data and applying deletions. Merge jobs are executed in parallel (in
+    /// multiple workers) and while other segments are being indexed. This restricts
+    /// us to only merge segments whose sequences are less than the smaller sequence
+    /// being indexed.
+    ///
+    /// As an example, if sequences 100 and 102 are indexed but 101 is still being
+    /// indexed, we can only merge segments with sequence <= 100. Otherwise, if we
+    /// merge 100 and 102 (generating a new 102 segment) and segment 101 included
+    /// deletions for 100, we'll never apply them and we'll end in an inconsistent
+    /// state.
+    pub async fn schedule_merges(&self, meta: &NidxMetadata, last_indexed_seq: Seq) -> anyhow::Result<()> {
+        let merges = sqlx::query!(
+            r#"
+            SELECT
+                index_id,
+                array_agg(
+                    (id, records)
+                    ORDER BY records DESC
+                ) AS "segments!: Vec<(SegmentId, i64)>"
+            FROM segments
+            WHERE
+                delete_at IS NULL AND merge_job_id IS NULL
+                AND seq <= $1
+                AND records <= $2
+            GROUP BY index_id"#,
+            i64::from(last_indexed_seq),
+            self.top_bucket_max_records as i64,
+        )
+        .fetch_all(&meta.pool)
+        .await?;
+
+        for m in merges {
+            let mut buckets = vec![];
+            let mut current_bucket = vec![];
+            let mut current_max_size_log = f64::MAX;
+
+            for (segment_id, records) in m.segments {
+                let segment_size_log = f64::from(std::cmp::max(records as u32, self.bottom_bucket_threshold)).log2();
+                if segment_size_log <= (current_max_size_log - self.bucket_size_log) {
+                    // traversed to next bucket, store current and continue
+                    buckets.push(current_bucket);
+                    current_bucket = vec![];
+                    current_max_size_log = segment_size_log;
+                }
+
+                current_bucket.push(segment_id);
+            }
+            buckets.push(current_bucket);
+
+            for segments in buckets {
+                if segments.len() >= self.min_number_of_segments {
+                    println!("Merge job for bucket: {segments:?}");
```

Comment:
Maybe switch to an info!/debug! line instead, e.g:

```suggestion
                    debug!(?segments, "Creating merge job for bucket");
```

Reply:
It was a leftover while debugging. I'll change it for a debug